src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -1208,14 +1252,44 @@ protected KeyRangeIterator execute(Executor executor)
             }
         }
 
+        @Override
+        protected boolean usesIncludedIndex()
+        {
+            return subplansSupplier.get().stream().anyMatch(KeysIteration::usesIncludedIndex);
+        }
+
         /**
          * Limits the number of intersected subplans
          */
         public Plan stripSubplans(int clauseLimit)
         {
             if (subplansSupplier.orig.size() <= clauseLimit)
                 return this;
-            List<Plan.KeysIteration> newSubplans = new ArrayList<>(subplansSupplier.orig.subList(0, clauseLimit));
+
+            if (factory.hints.included.isEmpty())
+                return withNewSubplans(new ArrayList<>(subplansSupplier.orig.subList(0, clauseLimit)));
+
+            List<Plan.KeysIteration> newSubplans = new ArrayList<>(clauseLimit);
+            List<Plan.KeysIteration> optionalSubplans = new ArrayList<>(clauseLimit);
```

Comment:
It will be more clear:
```suggestion
            List<Plan.KeysIteration> optionalNewSubplans = new ArrayList<>(clauseLimit);
```